tooling/templatize/pkg/pipeline/run.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		return nil, err
+	}
+
+	tmpl, err = tmpl.Parse(string(content))
+	if err != nil {
+		return nil, err
+	}
+
+	var tmplBytes bytes.Buffer
+	if err := tmpl.Option("missingkey=error").Execute(&tmplBytes, vars); err != nil {
+		return nil, err
+	}
+
+	pipeline := &Pipeline{}
+	err = yaml.Unmarshal(tmplBytes.Bytes(), pipeline)
+	if err != nil {
+		return nil, err
+	}
+	return pipeline, nil
+}
+
+func (p *Pipeline) Run(vars config.Variables) error {
+	for _, step := range p.Steps {
+		err := step.run(vars)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Step) run(vars config.Variables) error {
+	switch s.Action.Type {
+	case "Shell":
+		return s.runShell(vars)
+	default:
+		return fmt.Errorf("unsupported action type %q", s.Action.Type)
+	}
+}
+
+func (s *Step) runShell(vars config.Variables) error {
+	fmt.Printf("Running step %s: %q\n", s.Name, s.Action.Command)
+	fmt.Printf("Resource Group %s\n", s.ResourceGroup)
+	fmt.Printf("Subscription %s\n", s.Subscription)
+
+	// lookup subscription ID via name using the azure sdk
+	subscriptionID, err := lookupSubscriptionID(s.Subscription)
+	if err != nil {
+		return fmt.Errorf("failed to lookup subscription ID: %w", err)
+	}
+
+	// prepare kubeconfig
+	kubeconfigFile, err := buildAKSKubeConfig(
+		subscriptionID,
+		s.ResourceGroup,
+		"aro-hcp-aks",
+		context.Background(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to build kubeconfig: %w", err)
+	}
+
+	// schedule the deletion of the kubeconfig file after the command execution
+	defer func() {
+		if err := os.Remove(kubeconfigFile); err != nil {
+			fmt.Printf("Warning: failed to delete kubeconfig file %s: %v\n", kubeconfigFile, err)
+		}
+	}()
+
+	// build ENV vars
+	envVars := os.Environ()
+	envVars = append(envVars, fmt.Sprintf("KUBECONFIG=%s", kubeconfigFile))
+	for _, e := range s.Action.Variables {
+		value := vars[e.Value] // todo nested lookups
+		envVars = append(envVars, fmt.Sprintf("%s=%s", e.Name, value))
+	}
+
+	// execute the shell command with the environment variables
+	cmd := exec.Command("/bin/sh", "-c", s.Action.Command)
+	cmd.Dir = "/Users/goberlec/dev/aro-hcp/maestro/server"
+	cmd.Env = append(cmd.Env, envVars...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to execute shell command: %s %w", output, err)
+	}
+
+	// print the output of the command
+	fmt.Println(string(output))
+
+	return nil
+}
+
+func lookupSubscriptionID(subscriptionName string) (string, error) {
+	// Create a new Azure identity client
+	cred, err := azidentity.NewDefaultAzureCredential(nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to obtain a credential: %v", err)
+	}
+
+	// Create a new subscriptions client
+	client, err := armsubscriptions.NewClient(cred, nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to create subscriptions client: %v", err)
+	}
+
+	// List subscriptions and find the one with the matching name
+	ctx := context.Background()
```

Comment:
I would suggest creating a `ctx` child from `Background()` using `signal.NotifyContext()` and threading it through to these calls as well as your `exec.Cmd`. You'll be happy you did when you ctrl+C the CLI :)

https://pkg.go.dev/os/signal#NotifyContext

Reply:
done